Draft electric_plant_depreciation_functional_ferc1 transform.

zaneselvans opened the pull request:
# PR Overview

This is a draft transformer for FERC Form 1 Schedule 219 Section B: accumulated provision for depreciation of electric utility plant, broken down by plant functional classification (steam production, nuclear production, transmission, distribution, etc).

Currently the single data column in the final table is `utility_plant_value` which is the same as what was used in the `utility_plant_summary_ferc1` transformer (which this table aggregates into), but other related tables have used different data column names. We need to standardize across tables, but that's probably outside the scope of this PR.

This table reports the same total depreciated value as `electric_plant_depreciation_changes_ferc1` (Schedule 219 part A) but broken down into different categories. Currently there's a total category that's called `ending_balance_functional` which was convenient, but doesn't really feel right. Probably it should have the same name as the sister table that reports the same value as a total (associated with FERC Account 108).

This is one of 3 tables being generated based on the `f1_accumdepr_prvsn` DBF table. See: #1808 

# PR Checklist

- [ ] Merge the most recent version of the branch you are merging into (probably `dev`).
- [ ] All CI checks are passing. [Run tests locally to debug failures](https://catalystcoop-pudl.readthedocs.io/en/latest/dev/testing.html#running-tests-with-tox)
- [ ] Make sure you've included good docstrings.
- [ ] For major data coverage & analysis changes, [run data validation tests](https://catalystcoop-pudl.readthedocs.io/en/latest/dev/testing.html#data-validation)
- [ ] Include unit tests for new functions and classes.
- [ ] Defensive data quality/sanity checks in analyses & data processing functions.
- [ ] Update the [release notes](https://catalystcoop-pudl.readthedocs.io/en/latest/release_notes.html) and reference reference the PR and related issues.
- [ ] Do your own explanatory review of the PR to help the reviewer understand what's going on and identify issues preemptively.

cmgosnell commented on src/pudl/transform/ferc1.py:
```diff
@@ -336,12 +339,8 @@ class AlignRowNumbersDbf(TransformParams):
-) -> pd.DataFrame:
+def align_row_numbers_dbf(df: pd.DataFrame, params: AlignRowNumbersDbf) -> pd.DataFrame:
     """Rename the xbrl_factoid column after :meth:`align_row_numbers_dbf`."""
-    if params is None:
-        params = AlignRowNumbersDbf()
```
we've been doing this in many different way, which i would like to standardize. I am personally in favor of the method doing the work, so I like this.

cmgosnell commented on src/pudl/transform/ferc1.py:
```diff
@@ -3321,6 +3321,66 @@ def process_instant_xbrl(self, df: pd.DataFrame) -> pd.DataFrame:
+class ElectricPlantDepreciationFunctionalFerc1TableTransformer(
+    Ferc1AbstractTableTransformer
+):
+    """Transformer for :ref:`electric_plant_depreciation_functional_ferc1` table."""
```
ha yes we really should

cmgosnell commented on src/pudl/transform/params/ferc1.py:
```diff
@@ -3774,6 +3773,101 @@
+                    "accumulated_depreciation_distribution_ending_balance": "distribution_utility_plant_value",
+                    "accumulated_depreciation_general_ending_balance": "general_utility_plant_value",
+                    "accumulated_depreciation_regional_transmission_and_market_operation_ending_balance": "regional_transmission_and_market_operation_utility_plant_value",
+                    "accumulated_provision_for_depreciation_of_electric_utility_plant_ending_balance": "ending_balance_functional_utility_plant_value",
```
should it just be `total`?

cmgosnell commented on src/pudl/transform/params/ferc1.py:
```diff
@@ -3774,6 +3773,101 @@
+                }
+            },
+        },
+        "categorize_strings": {
+            "utility_type": UTILITY_TYPE_CATEGORIES,
+            "plant_classification_type": ELECTRIC_PLANT_CLASSIFICATION_CATEGORIES,
+        },
```
which is to say you renamed them above in `rename_columns_ferc1` so no need to apply `categorize_strings` after the rename -> wide_to_tidy reshape?

cmgosnell commented on src/pudl/metadata/fields.py:
```diff
@@ -1415,6 +1411,14 @@
         "description": "Net plant capability in megawatts.",
         "unit": "MW",
     },
+    "plant_classification_type": {
+        "type": "string",
+        "description": "Status of electric plant (in service, future, leased, total).",
+    },
+    "plant_functional_type": {
+        "type": "string",
+        "description": "Functional plant type (steam, hydro, distribution, etc.).",
+    },
```
can these be `plant_status` and `plant_type` or `asset_type` or something

cmgosnell replied:
can we rename these to `plant_status` and `plant_type` or `asset_type` or seomthing

cmgosnell commented on src/pudl/transform/params/ferc1.py:
```diff
@@ -3774,6 +3773,101 @@
+        "select_dbf_rows_by_category": {
+            "column_name": "plant_functional_type",
+            "select_by_xbrl_categories": False,
+            "additional_categories": [
+                "distribution",
+                "general",
+                "hydraulic_production_conventional",
+                "hydraulic_production_pumped_storage",
+                "nuclear_production",
+                "other_production",
+                "regional_transmission_and_market_operation",
+                "steam_production",
+                "transmission",
+                "ending_balance_functional",
+            ],
```
if you convert the `ending_balance_functional` to `total` and then add the `total` into `PLANT_FUNCTIONAL_CATEGORIES`, could you use the keys from `PLANT_FUNCTIONAL_CATEGORIES` instead of writing these out here?

cmgosnell reviewed (changes requested):
this looks pretty good! some naming requests and changes we discussed/that are laid out in your comments

zaneselvans commented:
@cmgosnell I think I've addressed all of the things we talked about. A couple of other things that came up:
* Added `report_id` in the `depreciation_amortization_summary_ferc1` table. I should check if any others are missing...
* Changed some logic in the top-level `pudl.transform.ferc1.transform()` function which didn't complain when I accidentally forgot to add the new table to the dictionary of tables.
* Removed the pudl_dtype application that was happening in the `transform()` function too, which dated from before we started using `enforce_schema()` on the tables.

zaneselvans commented on src/pudl/transform/ferc1.py:
```diff
@@ -3389,30 +3446,25 @@ def transform(
         "balance_sheet_assets_ferc1": BalanceSheetAssetsFerc1TableTransformer,
         "income_statement_ferc1": IncomeStatementFerc1TableTransformer,
         "electric_plant_depreciation_changes_ferc1": ElectricPlantDepreciationChangesFerc1TableTransformer,
+        "electric_plant_depreciation_functional_ferc1": ElectricPlantDepreciationFunctionalFerc1TableTransformer,
         "retained_earnings_ferc1": RetainedEarningsFerc1TableTransformer,
     }
-    # create an empty ditctionary to fill up through the transform fuctions
     ferc1_transformed_dfs = {}
-    # for each ferc table,
-    for table in ferc1_tfr_classes:
-        if table in ferc1_settings.tables:
-            logger.info(
-                f"Transforming raw FERC Form 1 dataframe for loading into {table}"
-            )
+    for table in ferc1_settings.tables:
+        if table == "plants_steam_ferc1":  # Special case, see below.
+            continue
+        logger.info(f"Transforming raw FERC Form 1 dataframe for loading into {table}")
```
This nested for loop was a temporary workaround from when we only had some of the original PUDL tables integrated, and it caused a silent failure when I added my new table to the ETL settings, but not to the dictionary of tables / functions listed above. Now if you give it settings for a table that has no transform it will fail, which I think is appropriate.

zaneselvans commented on src/pudl/transform/ferc1.py:
```diff
@@ -3429,11 +3481,7 @@ def transform(
             transformed_fuel=ferc1_transformed_dfs["fuel_ferc1"].copy(),
         )
 
-    # convert types and return:
-    return {
-        name: convert_cols_dtypes(df, data_source="ferc1")
-        for name, df in ferc1_transformed_dfs.items()
-    }
+    return ferc1_transformed_dfs
```
This is also a leftover from before we created the `enforce_schema()` step and integrated it into the parent transformer class, which made it redundant.

cmgosnell reviewed (approved):
do it 🟢 🎉 👻
